### DISCUSSION TITLE
Keeping critical files...

### DISCUSSION THREAD

**Discussion by KnifMelti** [Romanitho/Winget-AutoUpdate#135] [Ideas]
Could I make a PR with my solution to the importance of this?

Our scenario:
- We publish available apps in **SCCM/Software Center** for our users to self install/uninstall via **Winget-Install**
- **WhiteList** is used and **mods** for customizing
- NotificationLevel=None
- UpdatesInterval=Weekly
- WAUAutoUpdate on

The result is an **included_apps.txt** and **mods** directory that gets populated/unpopulated over time when users install/uninstall the published available apps, and it's different on each device.

Now we want to make a new deployment of **WAU** (better winget, incorporated own solution for running the System Task **Winget-AutoUpdate** via a shortcut in the Start Menu as a standard user and an integration with your Task **Winget-AutoUpdate-Notify** to make sure the user is aware of something happening when the shortcut is executed) but your **WAU-Uninstall.ps1** and **Winget-AutoUpdate-Install.ps1** deletes everything but the logs.

So my solution is to keep some files as they are, enabling us to upgrade **WAU** to our custom solution without interference.
Thus keeping all the **mods** and entries in the **included_apps.txt**.

Regarding https://github.com/Romanitho/Winget-AutoUpdate/discussions/108#discussioncomment-2964582 **"../because you may want to force a reinstall/..**" that could be solved by a **PARAMETER** to **Winget-AutoUpdate-Install.ps1**/**WAU-Uninstall.ps1** when installing/uninstalling (i.e. **-Clean** or **-NoClean** if  the default is set to **-Clean**).
The shift from **WhiteList** to **BlackList** and back doesn't get affected and **excluded_apps.txt** is always deleted.

BTW: "_../running the System Task **Winget-AutoUpdate** via a shortcut in the Start Menu as a standard user and an integration with your Task **Winget-AutoUpdate-Notify** to make sure the user is aware of something happening when the shortcut is executed/.._" could also be a native part of **WAU** :)

**Comment by Romanitho** [comment]
Very detailed :)

So, when WAU self update, it should keep all settings, config files and mods files.

When you uninstall/reinstall WAU, I agree, everything is removed except logs. It might be a good idea to implement a `-NoClean` settings (to keep default install/uninstall as it works today).

Regarding a shortcut to run WAU manually, I have it on my desktop :) but that is quite simple to implement and anybody can do it, as it is just a shortcut (and you can easily deploy it with SCCM or Intune). The thing is, the scheduled task is not readable for users as it is made to run as system. In your context, if I understand well, you've customized the scripts to run `Winget-AutoUpdate-Notify` when you want to run WAU manually, as your default notif level is "None". So I guess you have prepared dedicated notifications :)

**Reply by KnifMelti** [reply]
A little more intricate ;)...
You could:
When installling WAU:
```
[System.Diagnostics.EventLog]::CreateEventSource("Winget-AutoUpdate (WAU)", "Application")
[System.Diagnostics.EventLog]::WriteEntry("Winget-AutoUpdate (WAU)", "Winget-AutoUpdate (WAU) have been installed.", "Information", 1)
```
Set WAU to act on the Event:
![image](https://user-images.githubusercontent.com/102996177/181933580-d810dc80-c2eb-4ea0-8594-b78bc72bf1cc.png)

Create a shortcut that runs:
```
[System.Diagnostics.EventLog]::WriteEntry("Winget-AutoUpdate (WAU)", "Winget-AutoUpdate (WAU) will be started...", "Information", 100)
```
and calls `Winget-AutoUpdate-Notify` with dedicated notification...
Then a standard user can run it via the shortcut...

When uninstalling:
```
[System.Diagnostics.EventLog]::DeleteEventSource("Winget-AutoUpdate (WAU)")
```

**Reply by KnifMelti** [reply]
A classic trick to run something in System context via Task Scheduler (no GUI) - integrating with your Toast system...

**Reply by Romanitho** [reply]
I think we can also change rights to allow a system scheduled tasks to be executed from standard user (without write/edit rights).

**Reply by KnifMelti** [reply]
You really think so?
I'm not sure of that, but if so great!
Either way I think a registration of WAU in the event logs is a thing to consider...

**Reply by Romanitho** [reply]
that is as simple as this:

https://www.mysysadmintips.com/windows/clients/576-allow-to-view-and-run-scheduled-tasks-without-admin-rights

edit: not tried yet

**Reply by KnifMelti** [reply]
`Note: This method doesn't work on Windows 10 / Server 2016 and higher.`
In red, but we should try...

**Reply by KnifMelti** [reply]
Nope, doesn't work.. ..event log activation is the solution.

**Reply by Romanitho** [reply]
haha, very true. But I was not that free recently so not that time to test or add things in WAU :/

**Reply by Romanitho** [reply]
I'm back on this old post 😅
https://call4cloud.nl/2022/01/how-i-fell-in-love-with-app-updates/
apparently we can allow user to run system task. not tried on win10/11 yet

**Reply by KnifMelti** [reply]
CreateEventSource for WAU, set a trigger in the scheduled task on the event, make a shortcut to a script running in user context that writes an entry in WAU event log (thereby triggering the search for updates) and show a toast.
Easy/better/cleaner (just made a solution regarding mounting an ISO on a share only system has access to)...
...though his complicated way is good to know about!

**Reply by Romanitho** [reply]
ok, your solution seems good yes :D
I need to test that by curiosity.

**Reply by KnifMelti** [reply]
#171

**Reply by Romanitho** [reply]
I think I have it without event id and things, to keep that simplier 😋  
I've tried that and it seems to work great (just run this powershell as admin):  
```
$scheduler = New-Object -ComObject "Schedule.Service"
$scheduler.Connect()
$task = $scheduler.GetFolder("").GetTask("Winget-AutoUpdate")
$sec = $task.GetSecurityDescriptor(0xF)
$sec = $sec + '(A;;GRGX;;;AU)'
$task.SetSecurityDescriptor($sec, 0)
```
This way, Users have right to read and run the scheduled task without edit or delete rights.
Then, adding a shortcut on desktop as you did, and an option in the install script parameters to add the shortcut or no (and I think we can also allow WAU to be run manually only 😅)

**Reply by KnifMelti** [reply]
Then, disregard/end my PR.
Just keep commit f5824013a9822994b1b97d28a4c1d85b71e9e2e7

**Reply by KnifMelti** [reply]
You'll take it from here, then?

**Reply by KnifMelti** [reply]
Tried a bit... commit 8d6fd5bb3438d6c30e1619cf1c29fc7da51fdd19

**Reply by Romanitho** [reply]
Thanks for your work. I'll have a look in next days. Very busy atm :/

**Reply by Romanitho** [reply]
For the delta, I think it should be by default no ? I mean update system apps AND user apps, For example, if I run manually update check, it will run and update system apps (and list apps in a temp txt file) ans then, get avalable update apps that are not in the system list.
Any reason to have `-delta` ?

**Reply by KnifMelti** [reply]
This is **way** over my paygrade/PS knowledge ;)
Remember where I come from (cmd, just started with PS)...
...better then if you integrate the small pieces I got right in my PR, disregard/close it and take it from here!

**Reply by Romanitho** [reply]
You became a PS master ;)
I will PR my changes and you'll be able to add yours. Your work needs to come from you 🥰

**Comment by KnifMelti** [comment]
#136 #176 #178
